src/Microsoft.AspNet.Mvc.TagHelpers/Internal/FileVersionProvider.cs

```diff
@@ -0,0 +1,103 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Security.Cryptography;
+using Microsoft.AspNet.FileProviders;
+using Microsoft.AspNet.Hosting;
+using Microsoft.AspNet.WebUtilities;
+using Microsoft.Framework.Caching.Memory;
+using Microsoft.Framework.Runtime;
+
+namespace Microsoft.AspNet.Mvc.TagHelpers.Internal
+{
+    /// <summary>
+    /// Provides version hash for a specified file.
+    /// </summary>
+    public class FileVersionProvider
+    {
+        private const string versionKey = "v";
+
+        /// <summary>
+        /// Creates a new instance of <see cref="FileVersionProvider"/>.
+        /// </summary>
+        /// <param name="fileProvider">The file provider to get and watch files.</param>
+        /// <param name="applicationName">Name of the applicaiton.</param>
+        /// <param name="cache">Cache where versioned urls of files are cached.</param>
+        public FileVersionProvider(
+            IFileProvider fileProvider,
+            string applicationName,
+            IMemoryCache cache)
+        {
+            FileProvider = fileProvider;
+            ApplicationName = applicationName;
+            Cache = cache;
+        }
+
+        /// <summary>
+        /// The <see cref="IFileProvider"/> to get and watch files.
+        /// </summary>
+        public IFileProvider FileProvider { get; }
+
+        /// <summary>
+        /// The name of the application.
+        /// </summary>
+        public string ApplicationName { get; }
+
+        /// <summary>
+        /// The <see cref="IMemoryCache"/> to cache the versioned url of files.
+        /// </summary>
+        public IMemoryCache Cache { get; }
+
+        /// <summary>
+        /// Adds version query parameter to the specified file path.
+        /// </summary>
+        /// <param name="filePath">The path of the file to which version should be added.</param>
+        /// <returns>Path containing the version query string.</returns>
+        /// <remarks>
+        /// The version query string is appended as with the key "v".
+        /// </remarks>
+        public string AddVersionToFilePath(string filePath)
+        {
+            var fileInfo = FileProvider.GetFileInfo(filePath);
+            if (!fileInfo.Exists)
+            {
+                if (filePath.Contains(ApplicationName))
```

Comment:
You'll have to do a little bit more. 

- `Contains` covers `/` and `~/` starting cases but reallly leaves open situations when the `ApplicationName` is just randomly used in the URL (it can have false-positives which is realllly bad): Use `StartsWith`/`Substring`. This means you'll also need to strip out beginning `/` or `~/`.